lib/stdio/generator.js

```diff
@@ -0,0 +1,86 @@
+import {Duplex, Readable, PassThrough, getDefaultHighWaterMark} from 'node:stream';
+import {callbackify} from 'node:util';
+import {setTimeout} from 'node:timers/promises';
+
+// Generators can be used to transform/filter standard streams.
+// Generators have a simple syntax, yet allows all of the following:
+//  - Sharing state between chunks, by using logic before the `for` loop
+//  - Flushing logic, by using logic after the `for` loop
+//  - Asynchronous logic
+//  - Emitting multiple chunks from a single source chunk, even if spaced in time, by using multiple `yield`
+//  - Filtering, by using no `yield`
+// Therefore, there is no need to allow Node.js or web transform streams.
+// The `highWaterMark` is kept as the default value, since this is what `childProcess.std*` uses.
+// We ensure `objectMode` is `false` for better buffering.
+// Chunks are currently processed serially. We could add a `concurrency` option to parallelize in the future.
+export const generatorToTransformStream = ({value}, {encoding}) => {
+	const objectMode = false;
+	const highWaterMark = getDefaultHighWaterMark(objectMode);
+	const passThrough = new PassThrough({objectMode, highWaterMark, destroy: destroyPassThrough});
+	const iterable = passThrough.iterator();
+	const encodedIterable = applyEncoding(iterable, encoding);
+	const mappedIterable = value(encodedIterable);
+	const readableStream = Readable.from(mappedIterable, {objectMode, highWaterMark});
+	const duplexStream = Duplex.from({writable: passThrough, readable: readableStream});
+	return {value: duplexStream};
+};
+
+// When an error is thrown in a generator, the PassThrough is aborted.
+// This creates a race condition for which error is propagated, due to the Duplex throwing twice:
+//  - The writable side is aborted (PassThrough)
+//  - The readable side propagate the generator's error
+// In order for the later to win that race, we need to wait one microtask.
+const destroyPassThrough = callbackify(async error => {
+	await setTimeout(0);
```

Comment:
Shouldn't this use `queueMicrotask`?

Reply:
Waiting for one microtask (`queueMicrotask()`) works.

I am waiting for one macrotask instead (`setTimeout()`) to be on the safe side, in case the Node.js core logic changes in the future. The race condition this is solving relies on multiple core Node.js functions' implementation, specifically: 
  - [`Readable.from()`](https://github.com/nodejs/node/blob/e133e5115a829aa7a445cfd8754df3a3dd586b0a/lib/internal/streams/from.js#L200)'s error handling
  - [`readable.iterator()`](https://github.com/nodejs/node/blob/e133e5115a829aa7a445cfd8754df3a3dd586b0a/lib/internal/streams/readable.js#L1394-L1403)'s error handling
  - [stream destroyer](https://github.com/nodejs/node/blob/e133e5115a829aa7a445cfd8754df3a3dd586b0a/lib/internal/streams/destroy.js#L325)

The first two rely on `async`/`await`, so if core Node.js added an additional `await` in that logic, it is possible that it would impact that race condition.

To clarify the underlying issue: when the user-supplied generator throws, two errors are propagated in a race condition:
  - The `Duplex` writable part (the `PassThrough`) is aborted, i.e. throw an `AbortError` that is very generic and unhelpful for users
  - The `Duplex` readable part (wrapped by `Readable.from()`) propagated the original error

So we are delaying the first part by adding some delay to the `PassThrough`'s `destroy()` method. This delays destroying the generator's underlying `PassThrough` slightly longer, but that should not be an issue.

I did add a unit test for this specific case. However, I think we might want to be safe and wait for one macrotask instead of a microtask? What do you think?

Reply:
> In order for the later to win that race, we need to wait one microtask.

The code comment is incorrect then, since it says "microtask". I would expand a little bit on the comment (as succinctly as possible) there on why a macrosecond is used.